File: app/src/main/java/io/github/hathibelagal/mykanji/MainActivity.java
package io.github.hathibelagal.mykanji;

import android.annotation.SuppressLint;
import android.content.pm.ActivityInfo;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.ConsoleMessage;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.LinearLayout;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.OnApplyWindowInsetsListener;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import java.util.Objects;

public class MainActivity extends AppCompatActivity {

    @SuppressLint({"SetJavaScriptEnabled", "SourceLockedOrientationActivity"})
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setBackgroundColor(getResources().getColor(R.color.background));
        setContentView(layout);

        ViewCompat.setOnApplyWindowInsetsListener(layout, new OnApplyWindowInsetsListener() {
            @NonNull
            @Override
            public WindowInsetsCompat onApplyWindowInsets(@NonNull View v, @NonNull WindowInsetsCompat insets) {
                Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
                v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
                return insets;
            }
        });

        WebView wv = new WebView(this);
        WebSettings settings = wv.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setAllowFileAccess(true);
        settings.setDefaultFontSize(12);
        settings.setTextZoom(100);
        settings.setDatabaseEnabled(true);

        wv.setWebChromeClient(new WebChromeClient() {

        });

        wv.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

        wv.loadUrl("file:///android_asset/index.html");
        layout.addView(wv);

        if(getSupportActionBar() != null) {
            getSupportActionBar().hide();
        }
    }
}

File: app/src/main/java/io/github/hathibelagal/kanjitrainer/MainActivity.kt
package io.github.hathibelagal.kanjitrainer

import android.annotation.SuppressLint
import android.media.MediaPlayer
import android.os.Build
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.activity.enableEdgeToEdge
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

class MainActivity : AppCompatActivity() {
    private lateinit var kanjiItems: KanjiItems

    private val answer1: TextView by lazy { findViewById(R.id.answer_choice_1) }
    private val answer2: TextView by lazy { findViewById(R.id.answer_choice_2) }
    private val answer3: TextView by lazy { findViewById(R.id.answer_choice_3) }
    private val answer4: TextView by lazy { findViewById(R.id.answer_choice_4) }

    private lateinit var correctAnswer: String
    private lateinit var currentKanji: String
    private lateinit var vibrator: Vibrator

    private lateinit var correctSound: MediaPlayer
    private lateinit var incorrectSound: MediaPlayer

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        vibrator = getSystemService(VIBRATOR_SERVICE) as Vibrator

        kanjiItems = KanjiItems(this)
        kanjiItems.loadKanjiFromJSON()

        showKanji()

        answer1.setOnClickListener {
            isCorrectAnswer(answer1.text.toString())
        }
        answer2.setOnClickListener {
            isCorrectAnswer(answer2.text.toString())
        }
        answer3.setOnClickListener {
            isCorrectAnswer(answer3.text.toString())
        }
        answer4.setOnClickListener {
            isCorrectAnswer(answer4.text.toString())
        }

        correctSound = MediaPlayer.create(this, R.raw.correct)
        incorrectSound = MediaPlayer.create(this, R.raw.incorrect)

    }

    private fun playSound(correct: Boolean) {
        if(correct) {
            correctSound.seekTo(0)
            correctSound.start()
        } else {
            incorrectSound.seekTo(0)
            incorrectSound.start()
        }
    }

    private fun isCorrectAnswer(answer: String) {
        playSound(answer == correctAnswer)
        if(answer != correctAnswer) {
            kanjiItems.resetNSeen(currentKanji)
            vibrator.vibrate(400)
        }
        showKanji()
    }

    private fun showKanji() {
        val kanji = kanjiItems.getRandKanji()
        findViewById<TextView>(R.id.kanji_question).text = kanji.kanji
        val nSeenText = "Streak: ${kanji.nSeen + 1}"
        findViewById<TextView>(R.id.n_seen).text = nSeenText

        val meanings = kanjiItems.getRandomMeanings(kanji.kanji, 3)
        meanings.add(kanji.meaning)
        meanings.shuffle()

        correctAnswer = kanji.meaning
        currentKanji = kanji.kanji

        if(kanji.nSeen == 0) {
            answer1.text = kanji.meaning
            findViewById<LinearLayout>(R.id.other_answers).visibility = View.GONE
            findViewById<TextView>(R.id.answer_title).text = getString(R.string.correct_answer_is)
        } else {
            findViewById<TextView>(R.id.answer_title).text = getString(R.string.pick_an_answer)
            findViewById<LinearLayout>(R.id.other_answers).visibility = View.VISIBLE
            answer1.text = meanings[0]
            answer2.text = meanings[1]
            answer3.text = meanings[2]
            answer4.text = meanings[3]
        }

        kanjiItems.incrementNSeen(kanji.kanji)
    }
}